File: tools/common/gamedata_layout.py
"""Resolve where an install keeps its gamedata tables and paloc files.

Crimson Desert 2.01 renamed the `0008` gamedata directory and every one of
its file extensions, and split each language's single paloc blob into one
file per namespace:

    1.05 - 2.00   gamedata/binary__/client/bin/<table>.pabgb / .pabgh
                  gamedata/stringtable/binary__/localizationstring_<lang>.paloc
    2.01+         gamedata/binarystaticinfo__/bin/<table>.staticinfobody
                                                 /<table>.staticinfoheader
                  gamedata/stringtable/binary__/<lang>/<namespace>.paloc

The file *contents* did not change, so every parser is untouched and only
the lookup path moved. Resolution is newest-layout-first with a fallback,
so a kept pre-2.01 install still works.

Mirrors `vendor/crimson-rs/scripts/gamedata_layout.py` (upstream's own
test helper) and `src/CrimsonAtomtic.RustInterop/GameDataLayout.cs`.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

#: PAZ group holding the static-info gamedata tables.
GAMEDATA_GROUP = "0008"

#: Archive directory holding the localization files — or, since 2.01, the
#: per-language subdirectories holding them. The root itself did not move.
PALOC_ROOT = "gamedata/stringtable/binary__"

_PALOC_PREFIX = "localizationstring_"
_PALOC_SUFFIX = ".paloc"


@dataclass(frozen=True)
class BinLayout:
    """One naming scheme for the static-info tables."""

    dir: str
    body_ext: str
    header_ext: str

    def body(self, table_stem: str) -> str:
        """`"skill"` -> `skill.pabgb` / `skill.staticinfobody`."""
        return f"{table_stem}.{self.body_ext}"

    def header(self, table_stem: str) -> str:
        """`"skill"` -> `skill.pabgh` / `skill.staticinfoheader`."""
        return f"{table_stem}.{self.header_ext}"


#: Newest layout first, so a current install resolves on the first probe.
BIN_LAYOUTS: tuple[BinLayout, ...] = (
    BinLayout("gamedata/binarystaticinfo__/bin", "staticinfobody", "staticinfoheader"),
    BinLayout("gamedata/binary__/client/bin", "pabgb", "pabgh"),
)


def _read_pamt(game_dir: str | Path, group: str) -> dict | None:
    """Parse one group's manifest, or `None` when it isn't readable."""
    import crimson_rs  # type: ignore[import-not-found]

    pamt_path = Path(game_dir) / group / "0.pamt"
    if not pamt_path.is_file():
        return None
    try:
        return crimson_rs.parse_pamt_bytes(pamt_path.read_bytes())
    except Exception:
        return None


def _directory_names(pamt: dict) -> set[str]:
    return {d.get("path") or d.get("name") or "" for d in pamt["directories"]}


def resolve_bin_layout(
    game_dir: str | Path, group: str = GAMEDATA_GROUP
) -> BinLayout:
    """Pick the static-info layout this install actually ships.

    Raises `LookupError` when the manifest is unreadable or holds neither
    known layout — the latter means Pearl Abyss moved the tables again and
    `BIN_LAYOUTS` needs a new entry.
    """
    pamt = _read_pamt(game_dir, group)
    if pamt is None:
        raise LookupError(
            f"cannot read {Path(game_dir) / group / '0.pamt'} — "
            f"is the game installed at {game_dir}?"
        )
    present = _directory_names(pamt)
    for layout in BIN_LAYOUTS:
        if layout.dir in present:
            return layout
    raise LookupError(
        f"no known static-info directory in {group}/0.pamt; tried "
        + ", ".join(repr(x.dir) for x in BIN_LAYOUTS)
        + ". The gamedata tables moved again — add the new layout to "
        "BIN_LAYOUTS in tools/common/gamedata_layout.py."
    )


def paloc_files(
    game_dir: str | Path, group: str, lang: str
) -> tuple[str, tuple[str, ...]] | None:
    """Where one language's paloc file(s) live in `group`.

    Returns `(archive directory, filenames)` — one filename pre-2.01,
    many (sorted, so load order is stable) from 2.01 on. `None` when the
    group's manifest or the language is absent, which is the caller's cue
    to keep probing other groups.
    """
    pamt = _read_pamt(game_dir, group)
    if pamt is None:
        return None

    lang_dir = f"{PALOC_ROOT}/{lang}"
    for directory in pamt["directories"]:
        path = directory.get("path") or directory.get("name") or ""
        if path == lang_dir:
            # 2.01+: the directory itself is the language.
            names = tuple(sorted(
                f["name"] for f in directory["files"]
                if f["name"].endswith(_PALOC_SUFFIX)
            ))
            return (path, names) if names else None

    legacy_name = f"{_PALOC_PREFIX}{lang}{_PALOC_SUFFIX}"
    for directory in pamt["directories"]:
        path = directory.get("path") or directory.get("name") or ""
        if path != PALOC_ROOT:
            continue
        if any(f["name"] == legacy_name for f in directory["files"]):
            return (path, (legacy_name,))
    return None

File: src/CrimsonAtomtic.RustInterop/GameDataLayout.cs
namespace CrimsonAtomtic.RustInterop;

/// <summary>
/// Where the static-info gamedata tables and the localization blobs live
/// inside the live install's PAZ groups.
///
/// <para>
/// Crimson Desert 2.01 renamed the <c>0008</c> gamedata directory and every
/// one of its file extensions:
/// </para>
///
/// <code>
/// 1.05 - 2.00   gamedata/binary__/client/bin/&lt;table&gt;.pabgb / .pabgh
/// 2.01+         gamedata/binarystaticinfo__/bin/&lt;table&gt;.staticinfobody
///                                              /&lt;table&gt;.staticinfoheader
/// </code>
///
/// <para>
/// The file <i>contents</i> did not change — bodies and PABGH indices parse
/// byte-identically either way — so every crimson-rs parser is untouched and
/// only the lookup path moved. Mirrors
/// <c>vendor/crimson-rs/src/binary/gamedata_layout.rs</c>, which is
/// <c>#[cfg(test)]</c> over there and so not reachable through the C ABI.
/// </para>
///
/// <para>
/// Resolution is newest-layout-first with a fallback, so a pre-2.01 install
/// (the cross-version RE workflow keeps one around) still works.
/// </para>
/// </summary>
public sealed class GameDataLayout
{
    /// <summary>PAZ group holding the static-info gamedata tables.</summary>
    public const string GameDataGroup = "0008";

    /// <summary>
    /// Archive directory holding the localization files — or, since 2.01,
    /// the per-language subdirectories holding them. The root itself did
    /// not move.
    /// </summary>
    public const string PalocRoot = "gamedata/stringtable/binary__";

    private readonly string _bodyExtension;
    private readonly string _headerExtension;

    private GameDataLayout(string binDirectory, string bodyExtension, string headerExtension)
    {
        BinDirectory = binDirectory;
        _bodyExtension = bodyExtension;
        _headerExtension = headerExtension;
    }

    /// <summary>2.01+ layout. Probed first.</summary>
    public static GameDataLayout Modern { get; } =
        new("gamedata/binarystaticinfo__/bin", "staticinfobody", "staticinfoheader");

    /// <summary>1.05 – 2.00 layout.</summary>
    public static GameDataLayout Legacy { get; } =
        new("gamedata/binary__/client/bin", "pabgb", "pabgh");

    /// <summary>Newest layout first, so a current install resolves on the
    /// first probe and older ones fall through.</summary>
    private static readonly GameDataLayout[] ProbeOrder = [Modern, Legacy];

    /// <summary>
    /// Archive directory holding the static-info tables inside
    /// <see cref="GameDataGroup"/>.
    /// </summary>
    public string BinDirectory { get; }

    /// <summary>
    /// Table body filename: <c>"skill"</c> → <c>skill.pabgb</c> or
    /// <c>skill.staticinfobody</c>.
    /// </summary>
    public string Body(string tableStem) => $"{tableStem}.{_bodyExtension}";

    /// <summary>
    /// Table index filename: <c>"skill"</c> → <c>skill.pabgh</c> or
    /// <c>skill.staticinfoheader</c>.
    /// </summary>
    public string Header(string tableStem) => $"{tableStem}.{_headerExtension}";

    /// <summary>
    /// Resolve which layout an install ships by probing its
    /// <see cref="GameDataGroup"/> manifest for each candidate directory,
    /// newest first.
    /// </summary>
    /// <param name="paz">Extractor used to enumerate the manifest.</param>
    /// <param name="pamtPath">Absolute path to <c>0008/0.pamt</c>.</param>
    /// <returns>
    /// The layout the install ships. Reports <see cref="Modern"/> when the
    /// manifest is missing or unreadable — callers bail out on the absent
    /// install before the answer matters.
    /// </returns>
    public static GameDataLayout Resolve(IPazExtractor paz, string pamtPath)
    {
        ArgumentNullException.ThrowIfNull(paz);
        if (string.IsNullOrEmpty(pamtPath))
        {
            return Modern;
        }
        foreach (var candidate in ProbeOrder)
        {
            if (DirectoryExists(paz, pamtPath, candidate.BinDirectory))
            {
                return candidate;
            }
        }
        return Modern;
    }

    /// <summary>
    /// Every localization file for one language, as
    /// <c>(archive directory, filenames)</c>.
    ///
    /// <para>
    /// 2.01 split each language's single blob into one file per namespace
    /// inside a per-language subdirectory:
    /// </para>
    ///
    /// <code>
    /// 1.05 - 2.00   gamedata/stringtable/binary__/localizationstring_&lt;lang&gt;.paloc
    /// 2.01+         gamedata/stringtable/binary__/&lt;lang&gt;/&lt;namespace&gt;.paloc
    /// </code>
    ///
    /// <para>
    /// The namespace is already encoded in every entry's key, and the
    /// container is a flat entry list — so the split is presentational and
    /// the per-file catalogs can simply be queried in turn.
    /// </para>
    /// </summary>
    /// <returns>
    /// One filename pre-2.01, many from 2.01 on (sorted, so load order is
    /// stable). <c>null</c> when the group's manifest or the language is
    /// absent — the caller's cue to keep probing other groups.
    /// </returns>
    public static (string Directory, IReadOnlyList<string> Files)? ResolvePalocFiles(
        IPazExtractor paz, string pamtPath, string languageCode)
    {
        ArgumentNullException.ThrowIfNull(paz);
        if (string.IsNullOrEmpty(pamtPath) || string.IsNullOrEmpty(languageCode))
        {
            return null;
        }

        // 2.01+: the language IS the directory.
        var languageDirectory = $"{PalocRoot}/{languageCode}";
        var entries = TryListDir(paz, pamtPath, languageDirectory);
        if (entries is { Count: > 0 })
        {
            var names = entries
                .Select(e => e.Name)
                .Where(n => n.EndsWith(".paloc", StringComparison.OrdinalIgnoreCase))
                .OrderBy(n => n, StringComparer.Ordinal)
                .ToArray();
            if (names.Length > 0)
            {
                return (languageDirectory, names);
            }
        }

        // Pre-2.01: one blob per language, directly under the root.
        var legacyName = $"localizationstring_{languageCode}.paloc";
        var rootEntries = TryListDir(paz, pamtPath, PalocRoot);
        if (rootEntries is not null
            && rootEntries.Any(e => string.Equals(e.Name, legacyName, StringComparison.Ordinal)))
        {
            return (PalocRoot, (IReadOnlyList<string>)[legacyName]);
        }
        return null;
    }

    private static bool DirectoryExists(IPazExtractor paz, string pamtPath, string directory) =>
        TryListDir(paz, pamtPath, directory) is not null;

    /// <summary>
    /// <see cref="IPazExtractor.ListDir"/> with the "this install simply
    /// doesn't have it" outcomes folded into <c>null</c>. A missing
    /// directory is the normal answer while probing, not an error.
    /// </summary>
    private static IReadOnlyList<PazFileEntry>? TryListDir(
        IPazExtractor paz, string pamtPath, string directory)
    {
        try
        {
            return paz.ListDir(pamtPath, directory);
        }
        catch (CrimsonSaveException)
        {
            return null;
        }
        catch (IOException)
        {
            return null;
        }
    }
}
